### PR TITLE
Heroku-16 Support, Continued

### PR DESCRIPTION
**Pull request by iSDP**
Taking @pynixwang their edit and added required things for merging

closes #294

### CODE REVIEW

File: buildpacks/buildpack-clojure/buildpack-url [hunk 1]
```diff
@@ -1 +1 @@
-https://github.com/heroku/heroku-buildpack-clojure
+https://github.com/heroku/heroku-buildpack-clojure
```
Comment (michaelshobbs) [inline buildpacks/buildpack-clojure/buildpack-url]: Missing new line at eof. This might be related to the build issues.

Comment (josegonzalez) [reply buildpacks/buildpack-clojure/buildpack-url]: Weird that we'd need a newline.

Comment (iSDP) [reply buildpacks/buildpack-clojure/buildpack-url]: Newline was there before my edits, removed it to see if it was the issue, now added it agian, but the same end result remains....

```
#!/bin/bash -eo pipefail
tests=$(circleci tests glob buildpacks/*/tests/*/test.sh | circleci tests split --split-by=timings | xargs)
echo "executing tests: $tests"
basht $tests
Requested historical based timing, but they are not present.  Falling back to name based sorting
executing tests: buildpacks/buildpack-clojure/tests/clojure-ring/test.sh buildpacks/buildpack-java/tests/java-jetty/test.sh buildpacks/buildpack-play/tests/play/test.sh buildpacks/buildpack-scala/tests/scala/test.sh
=== RUN T_clojure-ring
::: BUILDING APP :::
       -----> Unable to select a buildpack
--- FAIL T_clojure-ring (1s)
=== RUN T_test_clojure-ring
-----> Unable to select a buildpack
--- FAIL T_test_clojure-ring (0s)
=== RUN T_java-jetty
::: BUILDING APP :::
       -----> Unable to select a buildpack
--- FAIL T_java-jetty (1s)
=== RUN T_play
::: BUILDING APP :::
       -----> Unable to select a buildpack
--- FAIL T_play (0s)
=== RUN T_scala
::: BUILDING APP :::
       -----> Unable to select a buildpack
--- FAIL T_scala (1s)

Ran 5 tests. 5 failed.

FAIL
Exited with code 5
```

Comment (iSDP) [reply buildpacks/buildpack-clojure/buildpack-url]: CircleCI passed before because `Dockerfile.dev` wasn't edited.

File: Dockerfile [hunk 2]
```diff
@@ -7,3 +8,4 @@ RUN /bin/herokuish buildpack install \
 	&& ln -s /bin/herokuish /exec
 COPY include/default_user.bash /tmp/default_user.bash
 RUN bash /tmp/default_user.bash && rm -f /tmp/default_user.bash
+RUN apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
```
Comment (michaelshobbs) [inline Dockerfile]: don't delete `/tmp/*`. that's where we bundle the buildpacks

https://github.com/gliderlabs/herokuish/blob/be0e139f66a7fa182715d79220fd1f4bc3140075/include/herokuish.bash#L13

Comment (iSDP) [reply Dockerfile]: ah dangit. that was @pynixwang's edit, i'll fix it.

**Comment by michaelshobbs** [comment]
Also needs a rebase

**Comment by michaelshobbs** [comment]
Also needs a rebase

File: buildpacks/buildpack-clojure/buildpack-url (hunk 1 above)
Comment (michaelshobbs) [reply buildpacks/buildpack-clojure/buildpack-url]: @josegonzalez we didn't but it's good practice and it was there before this PR

File: Dockerfile (hunk 2 above)
Comment (michaelshobbs) [reply Dockerfile]: Still need to change this line

File: Dockerfile [hunk 3]
```diff
@@ -1,9 +1,11 @@
-FROM heroku/cedar:14
-RUN curl "https://github.com/gliderlabs/herokuish/releases/download/v0.3.32/herokuish_0.3.32_linux_x86_64.tgz" \
+FROM heroku/heroku:16
+RUN apt-get update -qq && apt-get install -yy -q daemontools ruby-dev build-tools
+RUN curl "https://github.com/gliderlabs/herokuish/releases/download/v0.3.35/herokuish_0.3.35_linux_x86_64.tgz" \
```
Comment (michaelshobbs) [inline Dockerfile]: don't make this change yet. we handle that on release

File: Makefile [hunk 4]
```diff
@@ -1,6 +1,6 @@
 NAME = herokuish
 HARDWARE = $(shell uname -m)
-VERSION ?= 0.3.32
+VERSION ?= 0.3.35
```
Comment (michaelshobbs) [inline Makefile]: don't make this change yet. we handle that on release

Comment (iSDP) [reply Makefile]: I'll revert it, assumed it was needed because of the version defined in the Dockerfile

Comment (michaelshobbs) [reply Makefile]: `Dockerfile` is primarily used on push to `master` and `release` by Docker Hub to build the image for distribution. The new versioned artifact is created only on a successful CI run on the `release` branch.

**Comment by michaelshobbs** [comment]
Still needs rebase. We already had the php buildpack updated. Hence the new conflict on that file.

**Comment by michaelshobbs** [comment]
@iSDP I'd also like to see this squashed into a single commit. If you need help with that, let me know.

File: Dockerfile [hunk 5]
```diff
@@ -1,10 +1,11 @@
-FROM heroku/cedar:14
+FROM heroku/heroku:16
+RUN apt-get update -qq && apt-get install -yy -q daemontools ruby-dev build-essential
 RUN curl "https://github.com/gliderlabs/herokuish/releases/download/v0.3.35/herokuish_0.3.35_linux_x86_64.tgz" \
 		--silent -L | tar -xzC /bin
-RUN apt-get update && apt-get -qq -y --force-yes dist-upgrade && apt-get clean && rm -rf /var/cache/apt/archives/*
```
Comment (michaelshobbs) [inline Dockerfile]: we still want this line

Comment (iSDP) [reply Dockerfile]: in both dockerfiles?

Comment (michaelshobbs) [reply Dockerfile]: no. just the main `Dockerfile`

File: Dockerfile.dev [hunk 6]
```diff
@@ -1,8 +1,11 @@
-FROM heroku/cedar:14
+FROM heroku/heroku:16
+RUN apt-get update -qq && apt-get install -yy -q daemontools ruby-dev build-essential
 COPY ./build/linux/herokuish /bin/herokuish
+RUN apt-get update && apt-get -qq -y --force-yes dist-upgrade && apt-get clean && rm -rf /var/cache/apt/archives/*
```
Comment (michaelshobbs) [inline Dockerfile.dev]: let's not add this just yet

File: buildpacks/buildpack-clojure/tests/clojure-ring/Procfile [hunk 7]
```diff
@@ -1 +1 @@
-web: lein trampoline run
+web: lein trampoline run -m helloworld.web
```
Comment (michaelshobbs) [inline buildpacks/buildpack-clojure/tests/clojure-ring/Procfile]: @iSDP out of curiosity, why was this change needed?

Comment (iSDP) [reply buildpacks/buildpack-clojure/tests/clojure-ring/Procfile]: looked at the docs, showed that flag so added it.

File: Dockerfile (hunk 3 above)
Comment (iSDP) [reply Dockerfile]: damn, forgot to revert this i think

Comment (michaelshobbs) [reply Dockerfile]: 0.3.35 is the current version and this lines up with what is in master now. see current diff here: https://github.com/gliderlabs/herokuish/pull/333/files#diff-b67911656ef5d18c4ae36cb6741b7965

**Comment by michaelshobbs** [comment]
thinking we should release this as `0.4.0` and put a note at the top of the `README`. if folks want `cedar:14` they can build it themselves or pull `gliderlabs/herokuish:v0.3.5` manually. @josegonzalez thoughts

**Comment by iSDP** [comment]
CakePHP would be 816KB vs Symfony being just 76KB

For a simple test I think Symfony will suffice.

**Comment by pynixwang** [comment]
heroku-18 is coming soon.

**Comment by iSDP** [comment]
@pynixwang until then, heroku-16 is good enough.

**Comment by pynixwang** [comment]
heroku-16 lacks  many headers and libraries, can we use multistage build feature of docker to build with heroku-16-build and pack into heroku-16 image to run.

**Comment by iSDP** [comment]
the buildpacks provide the libraries.

**Comment by michaelshobbs** [comment]
@josegonzalez and i were discussing the build image in slack this morning. i just pushed that change to this branch and removed the `apt-get` install. according to this doc, we should be using the build version.

> The build image (heroku/heroku:16-build) is larger and includes development headers. It is only recommended for customers that need to compile source code.

https://devcenter.heroku.com/articles/heroku-16-stack#heroku-16-docker-image

**Comment by michaelshobbs** [comment]
Well that's weird....using the build image as the base, herokuish doesn't seem to detect any app types correctly even though executing `bin/detect` from an appropriate buildpack returns the same exit code and string as with the normal runtime image.....

**Comment by iSDP** [comment]
That's because heroku-16 uses heroku16-build as base.

**Comment by michaelshobbs** [comment]
@iSDP please explain further?

**Comment by michaelshobbs** [comment]
Oh wait. You have it backwards. `heroku:16-build` is based on `heroku:16`

https://github.com/heroku/stack-images/blob/master/heroku-16-build/Dockerfile

**Comment by iSDP** [comment]
Might as well integrate this then: https://github.com/heroku/stack-images/blob/master/heroku-16-build/bin/heroku-16-build.sh

**Comment by josegonzalez** [comment]
I think maybe they dont rebuild `heroku:16-build`? Maybe I read the comments wrong.

**Comment by iSDP** [comment]
Let's see how this holds up

**Comment by michaelshobbs** [comment]
I don't think maintaining our own copy of heroku's build script is the right answer.

**Comment by iSDP** [comment]
Not exactly maintaining our own copy, daemontools has been added as well, and this way we can customize the base image a bit.

**Comment by josegonzalez** [comment]
I think we should revert that and keep daemontools in our own Dockerfile.dev etc.

**Comment by michaelshobbs** [comment]
@josegonzalez revert which part exactly?

@iSDP the real solution here IMO is to figure out why building from `heroku/heroku:16-build` causes differing behavior in the buildpack `detect` scripts.

**Comment by iSDP** [comment]
Build image deletes /tmp in their bash script..

**Comment by michaelshobbs** [comment]
> Build image deletes /tmp in their bash script..

If our `Dockerfile` uses that image in the `FROM` then our command run *after* that deletion and thus our buildpacks would in our image....

**Comment by michaelshobbs** [comment]
I've built the image locally based on the `heroku/heroku:16-build` image. See below
```
 docker run -ti --rm herokuish:dev bash -c "ls -la /tmp/buildpacks"
total 60
drwxr-xr-x 15 root root 4096 Feb 15 17:15 .
drwxrwxrwt  1 root root 4096 Feb 15 17:15 ..
drwxr-xr-x  3 root root 4096 Feb 15 17:15 00_buildpack-multi
drwxr-xr-x  8 root root 4096 Feb 15 17:15 01_buildpack-ruby
drwxr-xr-x  9 root root 4096 Feb 15 17:15 02_buildpack-nodejs
drwxr-xr-x  9 root root 4096 Feb 15 17:15 03_buildpack-clojure
drwxr-xr-x  6 root root 4096 Feb 15 17:15 04_buildpack-python
drwxr-xr-x  9 root root 4096 Feb 15 17:15 05_buildpack-java
drwxr-xr-x  9 root root 4096 Feb 15 17:15 06_buildpack-gradle
drwxr-xr-x  8 root root 4096 Feb 15 17:15 07_buildpack-scala
drwxr-xr-x  6 root root 4096 Feb 15 17:15 08_buildpack-play
drwxr-xr-x  5 root root 4096 Feb 15 17:15 09_buildpack-php
drwxr-xr-x  6 root root 4096 Feb 15 17:15 10_buildpack-go
drwxr-xr-x  4 root root 4096 Feb 15 17:15 11_buildpack-erlang
drwxr-xr-x  4 root root 4096 Feb 15 17:15 12_buildpack-static
```

```
$ docker run -ti --rm -v ${PWD}/buildpacks/buildpack-nodejs/tests/nodejs-express:/tmp/app herokuish:dev bash -c "/bin/herokuish test / nodejs-express; /tmp/buildpacks/02_buildpack-nodejs/bin/detect /tmp/app; echo $?"
::: BUILDING APP :::
-----> Unable to select a buildpack
Node.js
0
```

**Comment by michaelshobbs** [comment]
Figured it out. buildpack detection was failing due to missing the `setuidgid` command.

**Comment by ArekCzarnik** [comment]
Any news to bring the pull request into master?

**Comment by josegonzalez** [comment]
@michaelshobbs remember this should be a 0.4.0 release ;)

**Comment by michaelshobbs** [comment]
I'm on it
